backend/ibutsu_server/controllers/login_controller.py

```diff
@@ -0,0 +1,134 @@
+    login = connexion.request.get_json()
+    if not login.get("email") or not login.get("password"):
+        return {"code": "EMPTY", "message": "Username and/or password are empty"}, 401
+    user = User.query.filter_by(email=login["email"]).first()
+    if user and user.check_password(login["password"]):
+        return {"name": user.name, "email": user.email, "token": generate_token(user.email)}
+    else:
+        return {"code": "INVALID", "message": "Username and/or password are invalid"}, 401
+
+
+def support():
+    """Return the authentication types that the server supports"""
+    return {
+        "user": True,
+        "keycloak": get_keycloak_config().get("client_id") is not None,
+        "google": get_provider_config("google")["client_id"] is not None,
+        "github": get_provider_config("github")["client_id"] is not None,
+        "facebook": get_provider_config("facebook")["client_id"] is not None,
+        "gitlab": get_provider_config("gitlab")["client_id"] is not None,
+    }
+
+
+def config(provider):
+    """Return the configuration for a particular login provider"""
+    if provider == "keycloak":
+        return get_keycloak_config(is_private=False)
+    else:
+        return get_provider_config(provider, is_private=False)
+
+
+def auth(provider):
+    """Auth redirect URL"""
+    if not connexion.request.args.get("code"):
+        return "Bad request", 400
+    code = connexion.request.args["code"]
+    frontend_url = build_url(
+        current_app.config.get("FRONTEND_URL", "http://localhost:3000"), "login"
+    )
+    if provider == "keycloak":
+        # Do the kc stuff
+        provider_config = get_keycloak_config(is_private=True)
+    else:
+        provider_config = get_provider_config(provider, is_private=True)
+    if provider == "google":
+        # Google does things its own way...
+        try:
+            # Verify the token
+            id_info = id_token.verify_oauth2_token(code, Request(), provider_config["client_id"])
+            # ID token is valid. Get the user's Google Account ID from the decoded token.
+            user = get_user_from_provider(provider, id_info)
+        except ValueError:
+            # Invalid token
+            return "Unauthorized", 401
+    else:
+        # For everyone else
+        payload = {
+            "client_id": provider_config["client_id"],
+            "code": code,
+            "grant_type": "authorization_code",
+            "redirect_uri": provider_config["redirect_uri"],
+        }
+        if provider_config.get("client_secret"):
+            payload["client_secret"] = provider_config["client_secret"]
+        response = requests.post(provider_config["token_url"], data=payload)
+        if response.status_code == 200:
+            if provider == "keycloak":
+                user = get_user_from_keycloak(response.json())
+            else:
+                user = get_user_from_provider(provider, response.json())
```

Comment:
If these requests fail, then the `user` will be undefined when we try to access the `id` in L121.

Reply:
OK, I changed this to catch no user and return a 401.